Aleix/run function calls sequentially

aconchillo opened the pull request:
#### Please describe the changes in your PR. If it is addressing an issue, please reference that as well.

See CHANGELOG.

filipi87 reviewed (commented):
Just to confirm before I start reviewing it. 

Based on the last meeting, this PR will be updated so that we can define through settings whether all function calls should run sequentially or in parallel, with the default behavior being parallel. Is that correct ?

aconchillo commented:
> Just to confirm before I start reviewing it.
> 
> Based on the last meeting, this PR will be updated so that we can define through settings whether all function calls should run sequentially or in parallel, with the default behavior being parallel. Is that correct ?

That is correct. This is now available. You can set `run_in_parallel=False` to run sequentially when you create the LLM service.

markbackman commented on CHANGELOG.md:
```diff
@@ -9,6 +9,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+- Function calls can now be executed sequentially (in the order received in the
+  completion) by passing `run_in_parallel=False` when creating your LLM
+  service. By default, if the LLM completion returns 2 or more function calls
+  they run concurrently. In both cases, concurrently and sequentailly, a new LLM
```
```suggestion
  they run concurrently. In both cases, concurrently and sequentially, a new LLM
```

aconchillo replied:
Fixed!

chadbailey59 reviewed (approved):
This looks good to me, pending Mark's one-character change; is there anything else I can help test?

markbackman commented on CHANGELOG.md:
```diff
@@ -73,6 +79,12 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Changed
 
+- Function calls are now cancelled by default if there's an interruption. To
```
I'm curious what changed your mind on this one.

markbackman replied:
I think that's fine. The good news is that it can be configured.

markbackman commented on src/pipecat/processors/aggregators/llm_response.py:
```diff
@@ -565,12 +565,14 @@ async def _handle_function_call_result(self, frame: FunctionCallResultFrame):
-                # If the tool call result has a run_llm property, use it
+                # If the tool call result has a run_llm property, use it.
                 run_llm = properties.run_llm
+            elif frame.run_llm is not None:
```
I'm wondering how this changes things. In flows, for function calls that transition nodes, we wait for function calls to run _and_ the subsequent node's context to be loaded before triggering a completion.

I think this change will break that.

aconchillo replied:
If I'm not mistaken, Flows uses `properties` and that is still working in the first if case. If Flows doesn't provide `properties.run_llm` whether it's True or False we will rely on the default behavior:

- Check properties.run_llm.
- If properties is not set, see if the result frames has a `run_llm`.
- If nothing was set, check if we are the last function call.

markbackman replied:
You're right! It uses:
```
@dataclass
class FunctionCallResultProperties:
    """Properties for a function call result frame."""

    run_llm: Optional[bool] = None
    on_context_updated: Optional[Callable[[], Awaitable[None]]] = None
```

markbackman commented on src/pipecat/services/llm_service.py:
```diff
@@ -42,8 +45,8 @@ async def __call__(
-class FunctionCallEntry:
-    """Represents an internal entry for a function call.
+class FunctionCallRegistryItem:
+    """Represents an entry of our function call registry.
```
This is one function call instance in the registry, right?

```suggestion
    """Represents an entry in our function call registry.
```

markbackman commented on src/pipecat/services/llm_service.py:
```diff
@@ -57,6 +60,46 @@ class FunctionCallEntry:
 
 
+@dataclass
+class FunctionCallLLM:
```
Maybe `FunctionCallToRegister` or something that better represents how it's used?

aconchillo replied:
Before I called them `FunctionCallRegistryItem`.

aconchillo replied:
Ah, sorry, not this one.

aconchillo replied:
This is actually a function call returned by the LLM.

aconchillo replied:
This should be the one that is called `FunctionCallToRun` because that's what the LLM wants to run.

aconchillo replied:
we ended up with `FunctionCallFromLLM`.

markbackman commented on src/pipecat/services/llm_service.py:
```diff
@@ -57,6 +60,46 @@ class FunctionCallEntry:
+
+
+@dataclass
+class FunctionCallRunner:
```
These are individual instances of function calls to run, right?

Calling it a Runner makes it sound like it's doing the running. Maybe `FunctionCallToRun`? (Naming is hard!)

aconchillo replied:
Before I called them `FunctionCallRunnerItem`. I even had a discussion with ChatGPT about the naming. Clearly that was not successful :joy:

aconchillo replied:
I will call it `FunctionCallRunnerItem` since we use the word item already

mattrossman commented:
When testing this PR locally, I still experience repeated responses when running a script with parallel function calls: [14-function-calling-parallel.py.zip](https://github.com/user-attachments/files/20375005/14-function-calling-parallel.py.zip) (see also [Discord context](https://discord.com/channels/1239284677165056021/1369046210857537698/1374518067463458836))

IIUC it is supposed to wait until all parallel functions complete and then generate a single response.

markbackman commented:
@mattrossman I see the same thing.

When run_in_parallel=True, I see:
- Function call 1
- LLM chat completion 1
- Function call 2
- LLM chat completion 2

When run_in_parallel=False, I see the expected behavior:
- Function call 1
- Function call 2
- LLM chat completion (for both 1 and 2)

markbackman commented on src/pipecat/services/openai_realtime_beta/openai.py:
```diff
@@ -574,25 +570,18 @@ async def _handle_assistant_output(self, output):
+            function_calls.append(
+                FunctionCallFromLLM(
+                    context=self._context,
+                    tool_call_id=item.tool_id,
```
```suggestion
                    tool_call_id=item.call_id,
```

aconchillo replied:
done!

markbackman commented on src/pipecat/services/openai_realtime_beta/openai.py:
```diff
@@ -574,25 +570,18 @@ async def _handle_assistant_output(self, output):
+                FunctionCallFromLLM(
+                    context=self._context,
+                    tool_call_id=item.tool_id,
+                    function_name=item.function_name,
```
```suggestion
                    function_name=item.name,
```

aconchillo replied:
done!

markbackman reviewed (commented):
This is looking much better! I'm now seeing OpenAI, Anthropic, Gemini, OpenAI Realtime, and Gemini Live all behave expectedly.

Two things I noticed:
- Running 14e results in Gemini saying "Let me check on that" once before the functions run (expected) and once again after they run. I don't see this for OpenAI.
- Running 14r (AWSBedrockLLMService) has an error when running a function call:
```
2025-05-31 09:45:04.133 | ERROR    | pipecat.services.aws.llm:_process_context:771 - AWSBedrockLLMService#0 exception: 'AWSBedrockLLMService' object has no attribute 'call_function'
Traceback (most recent call last):

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/examples/foundational/14r-function-calling-aws.py", line 164, in <module>
    main(run_example, transport_params=transport_params)
    │    │                             └ {'daily': <function <lambda> at 0x1568d7ba0>, 'twilio': <function <lambda> at 0x1560bd760>, 'webrtc': <function <lambda> at 0...
    │    └ <function run_example at 0x1560bd8a0>
    └ <function main at 0x1787c3740>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/examples/run.py", line 263, in main
    run_main(run_example, args, transport_params)
    │        │            │     └ {'daily': <function <lambda> at 0x1568d7ba0>, 'twilio': <function <lambda> at 0x1560bd760>, 'webrtc': <function <lambda> at 0...
    │        │            └ Namespace(host='localhost', port=7860, transport='webrtc', proxy=None, verbose=0)
    │        └ <function run_example at 0x1560bd8a0>
    └ <function run_main at 0x1787c36a0>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/examples/run.py", line 225, in run_main
    run_example_webrtc(run_example, args, params)
    │                  │            │     └ TransportParams(camera_in_enabled=False, camera_out_enabled=False, camera_out_is_live=False, camera_out_width=1024, camera_ou...
    │                  │            └ Namespace(host='localhost', port=7860, transport='webrtc', proxy=None, verbose=0)
    │                  └ <function run_example at 0x1560bd8a0>
    └ <function run_example_webrtc at 0x1787c3560>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/examples/run.py", line 149, in run_example_webrtc
    uvicorn.run(app, host=args.host, port=args.port)
    │       │   │         │    │          │    └ 7860
    │       │   │         │    │          └ Namespace(host='localhost', port=7860, transport='webrtc', proxy=None, verbose=0)
    │       │   │         │    └ 'localhost'
    │       │   │         └ Namespace(host='localhost', port=7860, transport='webrtc', proxy=None, verbose=0)
    │       │   └ <fastapi.applications.FastAPI object at 0x1572f9ac0>
    │       └ <function run at 0x157214f40>
    └ <module 'uvicorn' from '/Users/mbackman/src/daily-bots-sandbox/venv/lib/python3.12/site-packages/uvicorn/__init__.py'>

  File "/Users/mbackman/src/daily-bots-sandbox/venv/lib/python3.12/site-packages/uvicorn/main.py", line 579, in run
    server.run()
    │      └ <function Server.run at 0x157215120>
    └ <uvicorn.server.Server object at 0x1787c6450>
  File "/Users/mbackman/src/daily-bots-sandbox/venv/lib/python3.12/site-packages/uvicorn/server.py", line 65, in run
    return asyncio.run(self.serve(sockets=sockets))
           │       │   │    │             └ None
           │       │   │    └ <function Server.serve at 0x1572151c0>
           │       │   └ <uvicorn.server.Server object at 0x1787c6450>
           │       └ <function run at 0x103347a60>
           └ <module 'asyncio' from '/Users/mbackman/.pyenv/versions/3.12.8/lib/python3.12/asyncio/__init__.py'>
  File "/Users/mbackman/.pyenv/versions/3.12.8/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           │      │   └ <coroutine object Server.serve at 0x178773220>
           │      └ <function Runner.run at 0x1033a94e0>
           └ <asyncio.runners.Runner object at 0x103b7cdd0>
  File "/Users/mbackman/.pyenv/versions/3.12.8/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           │    │     │                  └ <Task pending name='Task-1' coro=<Server.serve() running at /Users/mbackman/src/daily-bots-sandbox/venv/lib/python3.12/site-p...
           │    │     └ <cyfunction Loop.run_until_complete at 0x103c49e50>
           │    └ <uvloop.Loop running=True closed=False debug=False>
           └ <asyncio.runners.Runner object at 0x103b7cdd0>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/utils/asyncio.py", line 107, in run_coroutine
    await coroutine
          └ <coroutine object FrameProcessor.__input_frame_task_handler at 0x30e0106d0>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/processors/frame_processor.py", line 371, in __input_frame_task_handler
    await self.process_frame(frame, direction)
          │    │             │      └ <FrameDirection.DOWNSTREAM: 1>
          │    │             └ OpenAILLMContextFrame(id=2073, name='OpenAILLMContextFrame#2', pts=None, metadata={}, transport_source=None, transport_destin...
          │    └ <function AWSBedrockLLMService.process_frame at 0x154572fc0>
          └ <pipecat.services.aws.llm.AWSBedrockLLMService object at 0x30024ecf0>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/services/aws/llm.py", line 807, in process_frame
    await self._process_context(context)
          │    │                └ <pipecat.services.aws.llm.AWSBedrockLLMContext object at 0x302ef8980>
          │    └ <function AWSBedrockLLMService._process_context at 0x154573060>
          └ <pipecat.services.aws.llm.AWSBedrockLLMService object at 0x30024ecf0>

  File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/utils/tracing/service_decorators.py", line 446, in wrapper
    result = await f(self, context, *args, **kwargs)
                   │ │     │         │       └ {}
                   │ │     │         └ ()
                   │ │     └ <pipecat.services.aws.llm.AWSBedrockLLMContext object at 0x302ef8980>
                   │ └ <pipecat.services.aws.llm.AWSBedrockLLMService object at 0x30024ecf0>
                   └ <function AWSBedrockLLMService._process_context at 0x154572f20>

> File "/Users/mbackman/src/daily-bots-sandbox/pipecat/src/pipecat/services/aws/llm.py", line 743, in _process_context
    await self.call_function(
          └ <pipecat.services.aws.llm.AWSBedrockLLMService object at 0x30024ecf0>

AttributeError: 'AWSBedrockLLMService' object has no attribute 'call_function'. Did you mean: 'has_function'?
```

aconchillo commented:
> This is looking much better! I'm now seeing OpenAI, Anthropic, Gemini, OpenAI Realtime, and Gemini Live all behave expectedly.
> 
> Two things I noticed:
> 
> * Running 14e results in Gemini saying "Let me check on that" once before the functions run (expected) and once again after they run. I don't see this for OpenAI.

Fixed. The issue is that `run_function_calls` was called with an empty list and the `on_function_calls_started` was called. But we just need to not run anything if not function calls is returned.

> * Running 14r (AWSBedrockLLMService) has an error when running a function call:
> 

Fixed. This was because this is a newer LLM service that at the time I started this PR was not there yet.
